diff --git a/hal/camera/camera/cfg_ftbl_custom_yuv_sub.h b/hal/camera/camera/cfg_ftbl_custom_yuv_sub.h
index ad1d8b5..5e6ba29 100755
--- a/hal/camera/camera/cfg_ftbl_custom_yuv_sub.h
+++ b/hal/camera/camera/cfg_ftbl_custom_yuv_sub.h
@@ -106,14 +106,13 @@ GETFINFO_YUV_SUB()
     )
 #endif
     //..........................................................................
-#if 0
-    //  Capture Size
+    //modify by zhou 0 to 1
+#if 1
+	//  Capture Size
+    //ylht modify for camera chazhi front camera
     CONFIG_FEATURE(FID_CAP_SIZE, 
         BY_DEFAULT(CAPTURE_SIZE_2560_1920), 
-        CAPTURE_SIZE_320_240,   CAPTURE_SIZE_640_480, 
-        CAPTURE_SIZE_1024_768,  CAPTURE_SIZE_1280_960, 
-        CAPTURE_SIZE_1600_1200, CAPTURE_SIZE_2048_1536, 
-        CAPTURE_SIZE_2560_1920
+        CAPTURE_SIZE_320_240,   CAPTURE_SIZE_640_480,  CAPTURE_SIZE_1024_768,  CAPTURE_SIZE_1280_960,  CAPTURE_SIZE_1600_1200, CAPTURE_SIZE_2048_1536, CAPTURE_SIZE_2560_1920
     )
 #endif
     //..........................................................................
